keps/sig-api-machinery/3157-watch-list/README.md

```diff
@@ -447,17 +447,35 @@ It replaces its internal store with the collected items (syncWith) and reuses th
 
 #### API changes
 
-Extend the optional `ResourceVersionMatch` query parameter of `ListOptions` with the following enumeration value:
+Extend the `ListOptions` struct with the following field:
 
 ```
-const (
-    // ResourceVersionMatchMostRecent matches data at the most recent ResourceVersion.
-    // The returned data is consistent, that is, served from etcd via a quorum read.
-    // For watch calls, it begins with synthetic "Added" events of all resources up to the most recent ResourceVersion.
-    // It ends with a synthetic "Bookmark" event containing the most recent ResourceVersion.
-    // For list calls, it has the same semantics as leaving ResourceVersion and ResourceVersionMatch unset.
-    ResourceVersionMatchMostRecent ResourceVersionMatch = "MostRecent"
-)
+type ListOptions struct {
+    ...
+
+    // SendInitialEvents, when set together with Watch option,
+    // being the watch stream with synthetic init events to build the
+    // whole state of all resources followed by a synthetic "Bookmark"
+    // event containing a ResourceVersion after which the server
+    // continues streaming events.
+    //
+    // When SendInitialEvents option is set, we require ResourceVersionMatch
+    // option to also be set. The semantic of the watch request is as following:
+    // - ResourceVersionMatch = NotOlderThan
+    //   It starts with sending initial events for all objects (at some resource
+    //   version), potentially followed by an event stream until the state
+    //   becomes synced to a resource version as fresh as the one provided by
+    //   the ResourceVersion option. At this point, a synthetic bookmark event
+    //   is send and watch stream is continued to be send.
+    //   If RV is unset, this is interpreted as "consistent read" and the
```

Comment:
I think this is be strange, setting `NotOlderThan without an RV` would give me a consistent read. At least for me, this would be unintuitive. I think in that case I would prefer to set `MostRecent with an RV`.

I think I'd forbid `NotOlderThan with an empty RV or an RV set to 0`

Reply:
I don't fully agree here.

1. RV="".
Unset RV for List operations is already used as "consistent read". So people should already be familiar with it.
However, here what we're saying is that you will get the state that is consistent, but the bookmark event that you will get will actually only satisfy the "NotOlderThan" semantic.

If you look at your implementation (and as we were discussing in the past, I don't think we can do anything better) what we're doing is:
- list from etcd to get the current RV (from which the consistent read is happening)
- once cache is synced to that value, start sending state

But given everything here is asynchronous, nothing really can't prevent you from the situation thatn when you get the RV from etcd, in the meantime something else happened and watchcache is already synced to the RV higher than what you get from etcd.
So we will effectively return bookmark at RV higher that what we though we should do.

So I think that this actually is inline with the semantic that we want to sell.


2. RV=0

This is potentially debatable. But note that my description doesn't mention RV=0 anywhere. It's treated as every other value of resource version - it's just that any RV is greater than 0, so state from any point in time satisfies that.
So forbidding RV=0, doesn't buy us anything here  - it neither won't change anything in the explanation above (so it will take the same effort for users to understand it whether we support it or not) nor it will change the implementation.

If we want to deprecate RV=0, we should approach that more holistically and do that across everywhere, not attack in in some places imho.

Reply:
I agree w/ Wojtek here also

Reply:
> So forbidding RV=0, doesn't buy us anything here

I agree. As I commented below, I'd actually go in the opposite direction and try to make rv=0 less of a special case... why not let someone watching from rv="" or rv=0 opt into getting bookmarks, or opt out of receiving initial events?